### PR TITLE
C51-53: Implement Bulk actions for case list

### PR DESCRIPTION
**Pull request by swastikpareek**
# Overview
This PR contains Bulk actions feature for case list page with new designs 

## Before 
c51-53 - before

## After
c51-53- bulk actions after

## Technical Details
* Created a reusable component `civicaseBulkActionsChecboxes` and `civicaseBulkActionsMessages` which uses the same controller.
* Every thing is handled by event subscribers and emitters. The following events are newly created
  * `civicase::bulk-actions::check-box-toggled` - triggers when bulk actions checkbox is clicked
  * `civicase::bulk-actions::bulk-selections ` - Fires when any bulk selection is clicked (Everything, visible, None)
  * `civicase::case-list::header-position-changed` - should be fired whenever the sticky case list headers position needs to be readjusted
  * `civicase::case-details::header-position-changed ` - should be fired whenever the sticky case summary tab header needs to readjusted.
  * `civicase::case-details::filter-position-changed` - should be fired whenever the sticky case detail filter header is to be readjusted.
 * Event subscribers are added in `caseListTable.js` for deselect/ select all/ select visible . Same can be referred for implementing bulk actions at other places.
* BEM classes used 
  * .civicase__bulkactions-checkbox
  * .civicase__bulkactions-checkbox-toggle
  * .civicase__bulkactions-select-mode-dropdown
  * .civicase__bulkactions-actions-dropdown 
  * .civicase__bulkactions-message 
  * .civicase__link-disabled - for disabling link with pointer-events set to none.
* Written tests for each part.

### Problem
When selecting a case and deselecting it again. the message box appears with 0 selections.

### Solution

 This was coming because the `bulkActionMessages` directive has `ng-hide` applied to itself in `caseListTable.html ` and had ng-show in its template. Since in the directive we were using `replace: true` both the conditions to the same div which sometimes incorrectly adds `ng-hide` class to the div.
`caseListTable.html`
```
<civicase-bulk-actions-message selected-items="selectedCases.length" total-count="totalCount" is-select-all-available="isSelectAllAvailable" show-checkboxes="showCheckboxes" ng-hide="viewingCase"></civicase-bulk-actions-message>
```
and `BulkActionsMessage.html`
```
<div class="civicase__bulkactions-message" ng-show="selectedItems > 0 && showCheckboxes">
```
in `BulkActionsMessage.js`
```
(function (angular, $, _) {
  var module = angular.module('civicase');

  module.directive('civicaseBulkActionsMessage', function () {
    return {
      restrict: 'EA',
      replace: true,
      controller: 'civicaseBulkActionsController',
      templateUrl: '~/civicase/BulkActionsMessage.html',
      scope: {
        selectedItems: '=',
        isSelectAllAvailable: '=',
        totalCount: '=',
        showCheckboxes: '='
      }
    };
  });
})(angular, CRM.$, CRM._);
```

So  removing `replace: true` from the directive fix this as now the ng-show from inside and ng-hide from outside will be applied on different div eliminating the chance of ambiguity. Updated the tests as well.

### CODE REVIEW

**Comment by deb1990** [comment]
2018-10-15 at 9 11 am

1. Actions text should not be ALL CAPS, also font color does not match design.
2. In Case Viewing mode, the Bulk Actions should be visible according to the design.

2018-10-15 at 9 13 am
As per discussion with Shitij, count of `Everything`, `Displayed` should be shown in brackets.

**Comment by deb1990** [comment]
2018-10-15 at 9 22 am
@swastikpareek how about having this row below the header of the table? could you check with Jon please.

File: ang/civicase/BulkActionsMessage.html [hunk 1]
```diff
@@ -0,0 +1,5 @@
+<div class="civicase__bulkactions-message" ng-show="selectedItems > 0 && showCheckboxes">
+  <div class="alert">
+    You have selected <strong>{{selectedItems}}</strong> <ng-pluralize count="selectedItems" when="{'one': 'item', 'other': 'items'}"></ng-pluralize>. <a ng-show="totalCount !== selectedItems" href="#" ng-class="{'civicase__link-disabled': !isSelectAllAvailable}" ng-click="select('all'); $event.stopPropagation(); $event.preventDefault();">Select all {{totalCount}} that match the search</a> <a ng-show="totalCount === selectedItems" href="#"  ng-class="{'civicase__link-disabled': !isSelectAllAvailable}" ng-click="select('none'); $event.stopPropagation(); $event.preventDefault();">Clear All</a>
```
Comment (deb1990) [inline ang/civicase/BulkActionsMessage.html]: Dont use `strong`, use `span` with css instead.

Comment (deb1990) [inline ang/civicase/BulkActionsMessage.html]: Please dont have such long lines, its hard to read. break them down.
```
<div class="civicase__bulkactions-message" ng-show="selectedItems > 0 && showCheckboxes">
  <div class="alert">
    You have selected <strong>{{selectedItems}}</strong>
    <ng-pluralize count="selectedItems" when="{'one': 'item', 'other': 'items'}"></ng-pluralize>.
    <a
      ng-show="totalCount !== selectedItems" href="#"
      ng-class="{'civicase__link-disabled': !isSelectAllAvailable}"
      ng-click="select('all'); $event.stopPropagation(); $event.preventDefault();">
      Select all {{totalCount}} that match the search
    </a>
    <a
      ng-show="totalCount === selectedItems" href="#"
      ng-class="{'civicase__link-disabled': !isSelectAllAvailable}"
      ng-click="select('none'); $event.stopPropagation(); $event.preventDefault();">
      Clear All
    </a>
  </div>
</div>
```

Comment (deb1990) [inline ang/civicase/BulkActionsMessage.html]: `$event.stopPropagation(); $event.preventDefault();` can be moved inside `select`.

File: ang/civicase/BulkActionsMessage.js [hunk 2]
```diff
@@ -0,0 +1,18 @@
+(function (angular, $, _) {
+  var module = angular.module('civicase');
+
+  module.directive('civicaseBulkActionsMessage', function () {
+    return {
+      restrict: 'EA',
+      replace: true,
+      controller: 'civicaseBulkActionsController',
```
Comment (deb1990) [inline ang/civicase/BulkActionsMessage.js]: Why dont you just use `civicaseBulkActionsController` as `ng-controller` in the HTML and remove the JS file. Would that work?

File: ang/civicase/CaseCard--case-list.html [hunk 3]
```diff
@@ -4,15 +4,15 @@
   * use 'civicase__animated-checkbox-card--expanded' class : for enabled checkbox state
   * Update links with activityFeedUrl function activityFeedUrl(caseid, activityType to link, activityStatus to Link)
   -->
-<div class="panel panel-default panel-strip-primary civicase__case-card civicase__animated-checkbox-card" ng-style="{'border-left-color': data.color}" ng-class="{'civicase__case-card--active': $parent.viewingCase === data.id, 'civicase__case-card--closed': data.status === 'Resolved'}" >
+<div class="panel panel-default panel-strip-primary civicase__case-card civicase__animated-checkbox-card" ng-style="{'border-left-color': data.color}" ng-class="{'civicase__case-card--active': $parent.viewingCase === data.id, 'civicase__case-card--closed': data.status === 'Resolved', 'civicase__animated-checkbox-card--expanded': $parent.showCheckboxes && !$parent.viewingCase && $parent.selectedCases && $parent.bulkAllowed}" >
```
Comment (deb1990) [inline ang/civicase/CaseCard--case-list.html]: Break long lines please.

File: ang/civicase/CaseCard--case-list.html [hunk 4]
```diff
@@ -4,15 +4,15 @@
...
   <div class="panel-body">
     <!--
       TODO: To be implemented when implementing bulk actions.
       Click event to be implemented as a part of bulk actions
      <div ng-if="!$parent.bulkAllowed" class="civicase__case-card-lock">
       <div class="civicase__case-card-lock__inner fa fa-lock"></div>
     </div> -->
-    <span class="civicase__checkbox" ng-if="!$parent.viewingCase && $parent.selectedCases && $parent.bulkAllowed" role="checkbox" aria-checked="{{activity.is_completed}}" tabindex="0" aria-labelledby="Mark complete/incomplete">
+    <span class="civicase__checkbox" ng-if="!$parent.viewingCase && $parent.selectedCases && $parent.bulkAllowed" role="checkbox" aria-checked="{{activity.is_completed}}" tabindex="0" aria-labelledby="Mark complete/incomplete" ng-click="toggleSelected(); $event.stopPropagation();">
       <i ng-if="data.selected" class="civicase__checkbox--checked material-icons">check_box</i>
```
Comment (deb1990) [inline ang/civicase/CaseCard--case-list.html]: Break long lines please.

File: ang/civicase/CaseCard--case-list.html [hunk 5]
```diff
@@ -4,15 +4,15 @@
   * use 'civicase__animated-checkbox-card--expanded' class : for enabled checkbox state
   * Update links with activityFeedUrl function activityFeedUrl(caseid, activityType to link, activityStatus to Link)
   -->
-<div class="panel panel-default panel-strip-primary civicase__case-card civicase__animated-checkbox-card" ng-style="{'border-left-color': data.color}" ng-class="{'civicase__case-card--active': $parent.viewingCase === data.id, 'civicase__case-card--closed': data.status === 'Resolved'}" >
+<div class="panel panel-default panel-strip-primary civicase__case-card civicase__animated-checkbox-card" ng-style="{'border-left-color': data.color}" ng-class="{'civicase__case-card--active': $parent.viewingCase === data.id, 'civicase__case-card--closed': data.status === 'Resolved', 'civicase__animated-checkbox-card--expanded': $parent.showCheckboxes && !$parent.viewingCase && $parent.selectedCases && $parent.bulkAllowed}" >
   <div class="panel-body">
     <!--
```
Comment (deb1990) [inline ang/civicase/CaseCard--case-list.html]: Do we still need this TODO?

File: ang/civicase/CaseCard--case-list.html [hunk 6]
```diff
@@ -4,15 +4,15 @@
   * use 'civicase__animated-checkbox-card--expanded' class : for enabled checkbox state
```
Comment (deb1990) [inline ang/civicase/CaseCard--case-list.html]: Do we still need this?

File: ang/civicase/CaseCard--case-list.html (hunk 3 above)
Comment (deb1990) [inline ang/civicase/CaseCard--case-list.html]: We should need to mention this many conditions.
`$parent.showCheckboxes && !$parent.viewingCase && $parent.selectedCases && $parent.bulkAllowed`

When `bulkAllowed` is false, `showCheckboxes` should be disabled by default. In fact the Bulk Allowed logic should be part of BulkActions directive. And unless user have permission, showCheckboxes should be disabled by default.

Also could you remind me, why `!$parent.viewingCase && $parent.selectedCases` is needed to be checked here?

File: ang/civicase/CaseCard.js [hunk 7]
```diff
@@ -15,12 +15,22 @@
...
     $scope.activityFeedUrl = getActivityFeedUrl;
     $scope.formatDate = DateHelper.formatDate;
     $scope.otherCategories = _.map(_.filter(CRM.civicase.activityCategories, function (category) {
       return category.name !== 'task' && category.name !== 'communication';
     }), function (category) {
       return category.name;
     });
+
+    /**
+     * Update the checkbox the case card
```
Comment (deb1990) [inline ang/civicase/CaseCard.js]: There are two kings of checkbox in the case card, we need to be more specific.
`Update Bulk Action checkbox of the case card`

File: ang/civicase/CaseCard.js [hunk 8]
```diff
@@ -15,12 +15,22 @@
   });
 
   module.controller('CivicaseCaseCardController', function ($scope, getActivityFeedUrl, DateHelper) {
+    $scope.data.selected = $scope.data.selected || false;
```
Comment (deb1990) [inline ang/civicase/CaseCard.js]: If `$scope.data.selected` is not true, doe not that mean its already false? Why is this line required?

File: ang/civicase/CaseCard.js [hunk 9]
```diff
@@ -15,12 +15,22 @@
...
     }), function (category) {
       return category.name;
     });
+
+    /**
+     * Update the checkbox the case card
+     */
+    $scope.toggleSelected = function () {
+      $scope.data.selected = !$scope.data.selected;
+      $scope.$emit('bulkSelectionCheckboxClicked', $scope.data);
```
Comment (deb1990) [inline ang/civicase/CaseCard.js]: Please follow this structure for emit messages.
`<appName>::<componentName/eventCategory>::<eventName>`
`'civicase::bulk-actions::check-box-toggled', $scope.activity);`

File: ang/civicase/CaseDetails.js [hunk 10]
```diff
@@ -311,12 +311,21 @@
...
-        var $casePanelBody = $('.civicase__case-details-panel > .panel-body');
-        var bodyPadding = parseInt($('body').css('padding-top'), 10); // to see the space for fixed menus
+      var $caseNavigation = $('.civicase__case-body_tab');
+      var $toolbarDrawer = $('#toolbar');
+      var $casePanelBody = $('.civicase__case-details-panel > .panel-body');
+      var bodyPadding = parseInt($('body').css('padding-top'), 10); // to see the space for fixed menus
+
+      $rootScope.$on('stickyCaseDetailsHeaderReinitiatePos', function () {
+        $timeout(function () {
+          // Resert right case view tab header
```
Comment (deb1990) [inline ang/civicase/CaseDetails.js]: Typo in `Resert`

File: ang/civicase/Search.js [hunk 11]
```diff
@@ -83,22 +83,8 @@
...
-        if ($tableHeader.data('bs.affix')) {
-          $tableHeader.data('bs.affix').options.offset.top = $tableHeader.offset().top - bodyPadding;
-        }
-
-        // Resert right case view tab header
-        if ($caseNavigation.data('bs.affix')) {
-          $caseNavigation.data('bs.affix').options.offset.top = $casePanelBody.offset().top - bodyPadding;
-        }
-      });
+      $rootScope.$broadcast('stickyCaseListHeaderReinitiatePos');
```
Comment (deb1990) [inline ang/civicase/Search.js]: Use `$emit`

Comment (deb1990) [reply ang/civicase/Search.js]: We dont need two events. Just fire a single event to denote, UI has changed. That should be listened by all.
For example, `civicase::case-list::header-position-changed`

File: ang/civicase/CaseListDirectives.js [hunk 12]
```diff
@@ -31,9 +31,32 @@
...
         scope.$watch('caseIsFocused', fixPositioningOnLoadingOrFocusChange);
-      }());
+      }
+
+      /**
+       * Initialise all subscribers
+       */
+      function initSubscribers () {
+        $rootScope.$on('stickyCaseListHeaderReinitiatePos', function () {
+          $timeout(function () {
```
Comment (deb1990) [inline ang/civicase/CaseListDirectives.js]: Why do we need `timeout` here?

File: ang/civicase/CaseListTable.html [hunk 13]
```diff
@@ -7,16 +8,16 @@
...
-              <input type="checkbox" ng-click="selectAll($event)" id="select-role-{{ $index }}" ng-checked="isSelection('all')" />
-              <label for="select-role-{{ $index }}"></label>
-            </span>
-            <span class="btn-group btn-group-sm" ng-if="!viewingCase && selectedCases && bulkAllowed && !item.lock">
-              <button type="button" class="btn btn-default dropdown-toggle" ng-class="{disabled: !selectedCases.length}" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
-                {{ ts('Actions') }}
-                <span class="caret"></span>
-              </button>
-              <ul class="dropdown-menu" civicase-actions="selectedCases" multiple="true" refresh-callback="refresh"></ul>
+            <!--  -->
```
Comment (deb1990) [inline ang/civicase/CaseListTable.html]: missing doc

File: ang/civicase/CaseListTable.html [hunk 14]
```diff
@@ -7,16 +8,16 @@
...
+            <civicase-bulk-actions-checkboxes ng-show="!viewingCase && selectedCases && bulkAllowed && !item.lock" class="pull-left" show-checkboxes="showCheckboxes" is-select-all-available="isSelectAllAvailable" selected-items="selectedCases.length"></civicase-bulk-actions-checkboxes>
```
Comment (deb1990) [inline ang/civicase/CaseListTable.html]: Break long lines.

File: ang/civicase/CaseListTable.html [hunk 15]
```diff
@@ -7,16 +8,16 @@
...
+            <span class="civicase__bulkactions-actions btn-group btn-group-sm" ng-if="!viewingCase && selectedCases && bulkAllowed && !item.lock">
```
Comment (deb1990) [inline ang/civicase/CaseListTable.html]: Same as the other comment
> We should need to mention this many conditions.
$parent.showCheckboxes && !$parent.viewingCase && $parent.selectedCases && $parent.bulkAllowed

> When bulkAllowed is false, showCheckboxes should be disabled by default. In fact the Bulk Allowed logic should be part of BulkActions directive. And unless user have permission, showCheckboxes should be disabled by default.

> Also could you remind me, why !$parent.viewingCase && $parent.selectedCases is needed to be checked here?

File: ang/civicase/CaseListTable.html [hunk 16]
```diff
@@ -7,16 +8,16 @@
...
+              <div class="btn-group">
+                <button type="button" class="btn btn-default" ng-class="{disabled: !selectedCases.length || !showCheckboxes}" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">{{ ts('Actions') }}</button>
+                <button type="button" class="btn btn-default dropdown-toggle" ng-class="{disabled: !selectedCases.length || !showCheckboxes}" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
+                  <span class="caret"></span>
+                </button>
+                <ul class="dropdown-menu" civicase-actions="selectedCases" multiple="true" refresh-callback="refresh"></ul>
+              </div>
```
Comment (deb1990) [inline ang/civicase/CaseListTable.html]: Extra space after after `div` should be removed.

File: ang/civicase/CaseListTable.js [hunk 17]
```diff
@@ -29,13 +30,98 @@
...
       getCases();
+      getAllCasesforSelectAll();
+      $scope.casePlaceholders = $scope.filters.id ? [0] : _.range($scope.page.size);
     }());
 
+    /**
+     * Initiate subscribers
+     */
+    function initSubscribers () {
+      $scope.$on('bulkSelection', function (event, condition) {
```
Comment (deb1990) [inline ang/civicase/CaseListTable.js]: Move the details to separate functions.
```
$scope.$on('bulkSelection', bulkSelectionEventListener);
$scope.$on('bulkSelectionCheckboxClicked', bulkSelectionCheckboxClickedListener);
```

File: ang/civicase/CaseListTable.js [hunk 18]
```diff
@@ -29,13 +30,98 @@
...
+        if (condition === 'none') {
+          deselectAllCases();
+        } else if (condition === 'visible') {
+          selectDisplayedCases();
+        } else if (condition === 'all') {
```
Comment (deb1990) [inline ang/civicase/CaseListTable.js]: `} else if (condition === 'all' && $scope.isSelectAllAvailable) {`

File: ang/civicase/CaseListTable.js [hunk 19]
```diff
@@ -29,13 +30,98 @@
...
+        } else {
+          _.remove($scope.selectedCases, {
+            id: data.id
+          });
+        }
+      });
+    }
+
+    /**
+     * Asynchronously get all cases for the select all bulk
```
Comment (deb1990) [inline ang/civicase/CaseListTable.js]: `Asynchronously get all cases for the bulk actions select all functionality `

File: ang/civicase/CaseListTable.js [hunk 20]
```diff
@@ -29,13 +30,98 @@
...
+      params[0][2].options.limit = 0;
+
+      crmApi(params).then(function (res) {
+        allCases = res[0].values;
+        $scope.isSelectAllAvailable = true;
+      });
+    }
+
+    /**
+     * Select all functionality
```
Comment (deb1990) [inline ang/civicase/CaseListTable.js]: `Select All Cases`

File: ang/civicase/CaseListTable.js [hunk 21]
```diff
@@ -29,13 +30,98 @@
...
+      $scope.selectedCases = [];
+    }
+
+    /**
+     * Select All visible data.
+     */
+    function selectDisplayedCases () {
+      _.each($scope.cases, function (item, index) {
+        $scope.cases[index].selected = true;
+        var isCurrentCaseInSelectedCases = _.find($scope.selectedCases, {
```
Comment (deb1990) [inline ang/civicase/CaseListTable.js]: Variable should be declared outside of loop.

File: ang/civicase/CaseListTable.js [hunk 22]
```diff
@@ -29,13 +30,98 @@
...
+        $scope.isSelectAllAvailable = true;
+      });
+    }
+
+    /**
+     * Select all functionality
+     */
+    function selectEveryCase () {
+      $scope.selectedCases = [];
+      $scope.selectedCases = _.each(allCases, formatCase);
```
Comment (deb1990) [inline ang/civicase/CaseListTable.js]: I though we wont be fomatting all cases, as its a performance problem. Why is it necessary?

File: ang/civicase/CaseListTable.js [hunk 23]
```diff
@@ -29,13 +30,98 @@
...
+      initSubscribers();
       getCases();
+      getAllCasesforSelectAll();
+      $scope.casePlaceholders = $scope.filters.id ? [0] : _.range($scope.page.size);
     }());
 
+    /**
+     * Initiate subscribers
+     */
+    function initSubscribers () {
```
Comment (deb1990) [inline ang/civicase/CaseListTable.js]: Private functions should be defined after public functions. JS refactoring wiki needs to be followed.

File: ang/civicase/CaseListTable.js [hunk 24]
```diff
@@ -210,13 +297,31 @@
...
+    }
+
+    /**
+     * Case Watcher - Updates the checkbox if a case is selected
+     *
+     * @params {array} cases
+     */
+    function casesWatcher (cases) {
+      // if case is in selectedCases array update the UI model (checkbox)
+      _.each(cases, function (item, index) {
```
Comment (deb1990) [inline ang/civicase/CaseListTable.js]: I think we discussed that we will be storing the IDs only. can we do that?

File: ang/civicase/CaseCard--case-list.html (hunk 6 above)
Comment (swastikpareek) [reply ang/civicase/CaseCard--case-list.html]: No we don't need this. Let me remove it.

File: ang/civicase/BulkActionsMessage.js (hunk 2 above)
Comment (swastikpareek) [reply ang/civicase/BulkActionsMessage.js]: This can't be done for following reasons
* This is a directive and if we want to make the HTML reusable then its better to use it as a directive. if we use the above approach then we will have to re write the whole HTML.
* Also the attributes that are used in this directive are specific to itself and make it reusable. For e.g. `selectedItems`  for cases will have different variable name (selected-items="selectedCases.length") and for activities might be different (selected-items="selectedActivities.length"

File: ang/civicase/CaseCard--case-list.html (hunk 3 above)
Comment (swastikpareek) [reply ang/civicase/CaseCard--case-list.html]: > When bulkAllowed is false, showCheckboxes should be disabled by default. In fact the Bulk Allowed logic should be part of BulkActions directive. And unless user have permission, showCheckboxes should be disabled by default.

`bulkAllowed` is for permissions and show checkbox is for toggle view, both are different. However, this is a nice idea to make BulkActions to be a part of the directive. Let me see to this. and get back to you.

> Also could you remind me, why !$parent.viewingCase && $parent.selectedCases is needed to be checked here?

Sure. In the current design the checkbox doesn't show up when any case is in view state also if at least one case is selected, that means checkbox were enabled and if user navigates within pagers then the checkboxes should be visible. Making user know that he/she have some selection already in progress. Also if one of the cases is selected the actions dropdown will be available and if it is then the user should see the checkboxes for each cases throughout pagers.

File: ang/civicase/CaseCard.js (hunk 8 above)
Comment (swastikpareek) [reply ang/civicase/CaseCard.js]: We needed this to initialise the variable as it was throwing some error for some condition. which I can't recall right now. However , removing this line don't show any impact on the functionality. So removing it.

File: ang/civicase/Search.js (hunk 11 above)
Comment (swastikpareek) [reply ang/civicase/Search.js]: > Use `$emit`

Limitation of using $emit is that it can be used in hierarchical way (i.e. any emitted event can only be caught by its parent scope or itself) but this event is being broadcasted from multiple controllers (Search.js) and (Bulkactionmessage.js) then we must need to use rootscope broadcast.

File: ang/civicase/CaseListTable.html (hunk 13 above)
Comment (swastikpareek) [reply ang/civicase/CaseListTable.html]: Its not missing doc but an empty inline comment missed . Let me remove it.

File: ang/civicase/CaseListTable.html (hunk 15 above)
Comment (swastikpareek) [reply ang/civicase/CaseListTable.html]: Please see my comments above

File: ang/civicase/CaseListTable.js (hunk 22 above)
Comment (swastikpareek) [reply ang/civicase/CaseListTable.js]: We need some keys like `item.status` that are created using formatCase Function. I don't think it is a big performance issue as this happens in parallel and is non blocking.

File: ang/civicase/CaseListTable.js (hunk 24 above)
Comment (swastikpareek) [reply ang/civicase/CaseListTable.js]: I think that we discussed it already that  we will be storing the full case object. As if we just store the keys, then we have to create new services for fetching the details of each case with an id and it will be a complex architecture and might introduce regressions.

File: ang/civicase/BulkActionsCheckboxes.js [hunk 25]
```diff
@@ -0,0 +1,44 @@
...
+        isSelectAllAvailable: '='
+      }
+    };
+  });
+
+  module.controller('civicaseBulkActionsController', function ($scope, $rootScope) {
+    $scope.showCheckboxes = false;
+
+    $scope.$watch('selectedItems', function () {
+      $rootScope.$broadcast('stickyCaseListHeaderReinitiatePos');
```
[the remark this answers is not in the document]
Comment (swastikpareek) [reply ang/civicase/BulkActionsCheckboxes.js]: We can't use $emit as its subscribers will only be available inside its parent hierarchy, but where it is subscribers are added, are outside that scope.
So if we emit `stickyCaseListHeaderReinitiatePos` from `BulkActionsCheckboxes` it can't be caught in CaseListDirective as it doesn't come in `BulkActionsCheckboxes` parent Hierarchy.

Let me know if we need discussion on it.

File: ang/civicase/CaseListDirectives.js (hunk 12 above)
Comment (swastikpareek) [reply ang/civicase/CaseListDirectives.js]: We have to call the logic when DOM has updated and the best way to ensure that is to call the logic after the angular is done with its DOM updates. To make sure it is called after that we need to add a timeout. 

Please let me know if we need to discuss it further.

**Comment by swastikpareek** [comment]
> 1.Actions text should not be ALL CAPS, also font color does not match design.

Fixed

> In Case Viewing mode, the Bulk Actions should be visible according to the design.

 We don't need to show the checkbox and the actions menu when the case is selected (Default behaviour of the old dcase site)

> As per discussion with Shitij, count of Everything, Displayed should be shown in brackets.

Added.

File: ang/civicase/BulkActionsCheckboxes.js [hunk 26]
```diff
@@ -0,0 +1,62 @@
...
+    };
+
+    /**
+     * Emits event for bulk selections
+     * Available event
+     * - 'all' : Select all that matches the search
+     * - 'visible' :  Selects all visible selections
+     * - 'none' : Deselects all
+     *
+     * @params {String} condition
```
Comment (deb1990) [inline ang/civicase/BulkActionsCheckboxes.js]: Params are not updated.

File: ang/civicase/BulkActionsCheckboxes.js (hunk 25 above)
[the remark this answers is not in the document]
Comment (deb1990) [reply ang/civicase/BulkActionsCheckboxes.js]: Yes, but if we always use `$rootScope.$watch` then this would work.

File: ang/civicase/BulkActionsMessage.html (hunk 1 above)
Comment (deb1990) [reply ang/civicase/BulkActionsMessage.html]: Sorry its fine to use `strong`.

File: ang/civicase/BulkActionsMessage.js (hunk 2 above)
Comment (deb1990) [reply ang/civicase/BulkActionsMessage.js]: Lets discuss this.

File: ang/civicase/CaseCard--case-list.html (hunk 3 above)
Comment (deb1990) [reply ang/civicase/CaseCard--case-list.html]: Lets discuss this.

File: ang/civicase/Search.js (hunk 11 above)
Comment (deb1990) [reply ang/civicase/Search.js]: Lets discuss this.

File: ang/civicase/CaseListDirectives.js (hunk 12 above)
Comment (deb1990) [reply ang/civicase/CaseListDirectives.js]: Lets discuss this.

File: ang/test/civicase/BulkActionsCheckboxes.spec.js [hunk 27]
```diff
@@ -0,0 +1,146 @@
...
+    beforeEach(function () {
+      originalEmitFunction = element.isolateScope().$emit;
+      element.isolateScope().$emit = jasmine.createSpy('$emit');
+    });
+
+    afterEach(function () {
+      element.isolateScope().$emit = originalEmitFunction;
+    });
+
+    describe('called for all selection', function () {
```
Comment (deb1990) [inline ang/test/civicase/BulkActionsCheckboxes.spec.js]: `when` prefix.

File: ang/test/civicase/BulkActionsCheckboxes.spec.js [hunk 28]
```diff
@@ -0,0 +1,146 @@
...
+      beforeEach(function () {
+        element.isolateScope().select('all');
+      });
+
+      it('emits bulkSelection event with "all" parameter', function () {
+        expect(element.isolateScope().$emit).toHaveBeenCalledWith('bulkSelection', 'all');
```
Comment (deb1990) [inline ang/test/civicase/BulkActionsCheckboxes.spec.js]: The event name should follow the format, as per the previous comment.

File: ang/test/civicase/BulkActionsCheckboxes.spec.js [hunk 29]
```diff
@@ -0,0 +1,146 @@
...
+      });
+
+      it('emits bulkSelection event with "none" parameter', function () {
+        expect(element.isolateScope().$emit).toHaveBeenCalledWith('bulkSelection', 'none');
+      });
+    });
+  });
+
+  /**
+   * Function responsible for setting up compilation of the directive
```
Comment (deb1990) [inline ang/test/civicase/BulkActionsCheckboxes.spec.js]: `Compile the directive`

File: scss/components/_bulk-actions.scss [hunk 30]
```diff
@@ -0,0 +1,71 @@
...
+
+  .civicase__checkbox--checked--hide {
+    color: $crm-grayblue-darker;
+    font-size: 19.5px;
+    left: 3px;
+    top: 2px;
+  }
+}
+
+.civicase__bulkactions-checkbox-dropdown {
```
Comment (deb1990) [inline scss/components/_bulk-actions.scss]: `civicase__bulkactions-select-mode-dropdown`

File: scss/components/_bulk-actions.scss [hunk 31]
```diff
@@ -0,0 +1,71 @@
...
+  background: $crm-white;
+  padding: 4px 5px;
+  vertical-align: middle;
+}
+
+.civicase__bulkactions-actions {
```
Comment (deb1990) [inline scss/components/_bulk-actions.scss]: `civicase__bulkactions-actions-dropdown`

File: ang/test/civicase/BulkActionsMessage.spec.js [hunk 32]
```diff
@@ -0,0 +1,143 @@
+/* eslint-env jasmine */
+
+describe('BulkActionsMessage', function () {
```
[the remark this answers is not in the document]
Comment (swastikpareek) [reply ang/test/civicase/BulkActionsMessage.spec.js]: We are checking basic tests and the logical implementation inside the HTML template.

**Comment by swastikpareek** [comment]
@deb1990 : We do need timeouts in the subscribers for `header-position` subscribers. The top offset calculations is done correctly when we use timeouts.
